host/host.go

```diff
@@ -68,24 +68,79 @@ type (
+	// A SettingsReporter returns the host's current settings.
+	SettingsReporter interface {
+		Settings() rhp.HostSettings
+	}
+
+	// A TransactionPool broadcasts transaction sets to miners for inclusion in
+	// an upcoming block.
+	TransactionPool interface {
+		AddTransaction(txn types.Transaction) error
+		RecommendedFee() types.Currency
+		Transactions() []types.Transaction
+	}
+
+	// A Wallet provides addresses and funds and signs transactions.
+	Wallet interface {
+		Address() types.Address
+		FundTransaction(txn *types.Transaction, amount types.Currency, pool []types.Transaction) ([]types.ElementID, func(), error)
+		SignTransaction(vc consensus.ValidationContext, txn *types.Transaction, toSign []types.ElementID) error
+	}
+)
+
+type (
+	// ContractState is the current lifecycle stage of a contract.
+	ContractState string
+
+	// A Contract contains metadata on the current lifecycle stage of a file
+	// contract.
+	Contract struct {
+		rhp.Contract
+		Parent types.FileContractElement
+		// FormationTransaction is the transaction created by the host and
+		// renter during contract formation. A reference is kept in case it
+		// needs to be rebroadcast. Transaction proofs should be updated by the
+		// contract store if the transaction is not confirmed.
+		Confirmed            bool
+		FormationTransaction types.Transaction
+
+		// StorageProof is the future storage proof for the contract.
+		// WindowStart and WindowProof are expected to be kept updated by the
+		// contract store.
+		StorageProof types.StorageProof
+
+		// ResolutionHeight is the height the contract was resolved, or 0 if the
+		// contract is unresolved.
+		ResolutionHeight uint64
+		// State is the current lifecycle state of the contract.
+		State ContractState
+	}
+
+	// A ContractStore stores contracts and manages proofs for the host.
 	ContractStore interface {
 		chain.Subscriber
 
 		// Exists returns true if the contract is in the store.
 		Exists(types.ElementID) bool
 		// Get returns the contract with the given ID.
 		Get(types.ElementID) (rhp.Contract, error)
-		// Add stores the provided contract, overwriting any previous contract
-		// with the same ID.
+		// Add stores the provided contract, should error if the contract
+		// already exists in the store.
 		Add(rhp.Contract, types.Transaction) error
+		// Delete removes the contract with the given ID from the store.
+		Delete(types.ElementID) error
 		// ReviseContract updates the current revision associated with a contract.
 		Revise(rhp.Contract) error
 
 		// Roots returns the roots of all sectors stored by the contract.
 		Roots(types.ElementID) ([]types.Hash256, error)
 		// SetRoots sets the stored roots of the contract.
 		SetRoots(types.ElementID, []types.Hash256) error
+
+		// ContractAction calls contractFn on every contract in the store that
+		// needs a lifecycle action performed.
+		ContractAction(vc consensus.ValidationContext, contractFn func(consensus.ValidationContext, Contract)) error
```

Comment:
seems weird to pass a `vc` which is then immediately passed back into the `contractFn` you supplied. It looks like `vc` is used to decide which contracts are actionable; maybe instead, we could do:
```go
ForEachContract(fn func(Contract)) error
```
and have `fn` be responsible for skipping over contracts that don't require any action.

Reply:
Having the manager loop through each contract in the store feels bad from a scalability standpoint. Maybe it's best to use the index from the store instead of passing it in?

Reply:
true. Maybe we can get away with `ContractAction(height uint64, fn func(Contract)) error`? Then make `fn` a closure around the current `vc`.